### PR TITLE
[Logging.EventSource] Add trace correlation fields

### LINKED ISSUE #76718: EventSourceLogger does not propagate SpanId and TraceId
### Description

Asp.net log scope data collected from an out of process collector such as dotnet-monitor no longer contains Activity information such as TraceId and SpanId. These properties are no longer added as part of the Request scope data (see https://github.com/dotnet/aspnetcore/pull/22376). These properties are no added to loggers using IExternalScopeProvider/ISupportExternalScope, which EventSourceLogger does not support.

Related dotnet-monitor issue https://github.com/dotnet/dotnet-monitor/issues/2545

### Reproduction Steps

Create a new asp.net core 6.0 web application.
Install dotnet-monitor tool
Run `dotnet-monitor collect --no-auth`
Run `curl https://localhost:52323/logs?level=Information&pid=<pid of web app>`
Issue a request to the web app.

### Expected behavior

Log from dotnet-monitor should contain SpanId and TraceId information

### Actual behavior

Log entry contains RequestId and RequestPath, but no activity information is displayed.

### Regression?

Yes; this behavior worked in 3.1

### Known Workarounds

_No response_

### Configuration

Windows .Net 6 x64

The issue is specific to net 6+. 

### Other information

_No response_

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #76718

## Changes

* Adds `ActivityTraceId`, `ActivitySpanId`, and `ActivityTraceFlags` from `Activity.Current` to log events

/cc @noahfalk @tarekgh

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs [hunk 1]
```diff
@@ -128,50 +128,81 @@ private LoggingEventSource() : base(EventSourceSettings.EtwSelfDescribingEventFo
...
+            string? ActivityTraceId,
+            string? ActivitySpanId,
+            string? ActivityTraceFlags)
         {
-            if (IsEnabled())
+            Debug.Assert(LoggerName != null);
+
+            EventName ??= string.Empty;
+            FormattedMessage ??= string.Empty;
+            ActivityTraceId ??= string.Empty;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:149]: Should we just pass Empty instead of passing null and then recheck? this apply to other added activity properties.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs [hunk 2]
```diff
@@ -505,6 +554,8 @@ private static unsafe void SetEventData<T>(ref EventData eventData, ref T value,
             {
                 string str = (value as string)!;
 #if DEBUG
+                Debug.Assert(str != null);
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:557]: why did you remove this assert?

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs [hunk 3]
```diff
@@ -57,23 +58,58 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
...
+            {
+                activityTraceId = activity.TraceId.ToHexString();
+                activitySpanId = activity.SpanId.ToHexString();
+                activityTraceFlags = activity.ActivityTraceFlags == ActivityTraceFlags.None
+                    ? "0"
+                    : "1";
+            }
+            else
+            {
+                activityTraceId = null;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:89]: I added another comment that we can intialize these with empty string and avoid checking again in the event emission.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs [hunk 4]
```diff
@@ -57,23 +58,58 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
...
+                return;
+            }
+
             string? message = null;
 
+            Activity? activity = Activity.Current;
+            string? activityTraceId;
+            string? activitySpanId;
+            string? activityTraceFlags;
+            if (activity != null && activity.IdFormat == ActivityIdFormat.W3C)
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:79]: Are we sure we don't want support non W3C cases too?

File: src/libraries/Microsoft.Extensions.Logging.EventSource/tests/EventSourceLoggerTest.cs [hunk 5]
```diff
@@ -195,6 +195,67 @@ public void Logs_AsExpected_WithDefaults()
...
+                using Activity activity = new Activity("TestOperation");
+
+                if (useW3CFormatId)
+                {
+                    activity.SetIdFormat(ActivityIdFormat.W3C);
+                }
+                else
+                {
+                    activity.SetIdFormat(ActivityIdFormat.Hierarchical);
+                }
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/tests/EventSourceLoggerTest.cs:225]: add these and `sampled` inside the `hasTrace` check?

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:79]: Well for non-W3C TraceId is going to spit out `00000000000000000000000000000000` and SpanId is going to spit out `0000000000000000`. We could just push `Activity.Id` into the events which would include `W3C`, `Hierarchical`, and (presumably) anything added in the future. BUT. Accessing `Activity.Id` will force an allocation in the process being monitored. And it will force consumers to have to inspect the value and understand the possible formats/differences. Probably also the consumers will have to chop up the `Id` into the individual components. In my world of OTel I don't need hierarchical but if you feel there is a need and the tradeoffs are OK, I'm happy to switch it.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/tests/EventSourceLoggerTest.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/tests/EventSourceLoggerTest.cs]: Updated

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs]: Updated

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs]: Updated

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:557]: I added it 😄

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:557]: Oh, somehow the diff in the tool came red. or I was dreaming :-)

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs (hunk 4 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:79]: We have been supporting the `Hierarchical` cases with external logger scopes. 

https://github.com/dotnet/runtime/blob/b93061adc14eaaf5e83e366f9fba507ad917308c/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs#L291

Also, will it make sense to add the parent Id to the event as we do in scopes? 

CC @noahfalk if he has more insight if we should ignore the `Hierarchical` cases or not.

Comment (CodeBlanch) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:79]: Just FYI the reason I went with `ActivityTraceId`, `ActivitySpanId`, and `ActivityTraceFlags` is because those are the only things defined on the OTel Logs Data Model for correlation: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition

Happy to add other things, but from OTel perspective they aren't required.

Comment (noahfalk) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLogger.cs:79]: I'd be fine omitting support for hierarichal. Worst case I forsee is we get feedback that someone wants Hierarichal support, we decide the scenario in the request is important and then we add that too in a future release. I think that has a pretty low chance of happening.

**Review by tarekgh — APPROVED** [review]
Let's wait @noahfalk approval and then we can proceed merging.

**Review by noahfalk — APPROVED** [review]
LGTM 👍

**Comment by tarekgh** [comment]
@CodeBlanch I am seeing test failure in https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-103655-merge-e2f8c48b4c15440fa9/Microsoft.Extensions.Logging.EventSource.Tests/1/console.3e8ced58.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-103655-merge-e2f8c48b4c15440fa9/Microsoft.Extensions.Logging.EventSource.Tests/1/console.3e8ced58.log?helixlogtype=result) which can be related.
